Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs

```diff
@@ -0,0 +1,1791 @@
+                null,
+                @event.ContainingType,
+                false);
+            currentTypeDeclaration.AddMethodReference(delegateSymbolType, removeMethodReference);
+        }
+
+        void AddProjectedType(INamedTypeSymbol type)
+        {
+            currentTypeDeclaration = new TypeDeclaration(type);
+
+            foreach (var member in type.GetMembers())
+            {
+                if (member is IMethodSymbol method && 
+                    (method.MethodKind == MethodKind.Ordinary || method.MethodKind == MethodKind.Constructor))
+                {
+                    AddMethodReference(method);
+                }
+                else if (member is IPropertySymbol property)
+                {
+                    AddPropertyReference(property);
+                }
+                else if (member is IEventSymbol @event)
+                {
+                    AddEventReference(@event);
+                }
+                else
+                {
+                    Logger.Log("member not recognized: " + member.Kind + " " + member.Name);
+                }
+            }
+
+            typeDefinitionMapping[type.ToString()] = currentTypeDeclaration;
+        }
+
+        enum SynthesizedInterfaceType
+        {
+            Static,
+            Factory,
+            Default
+        }
+
+        string GetSynthesizedInterfaceName(string className, SynthesizedInterfaceType type)
+        {
+            // TODO: handle existing types by appending number suffix
+            return "I" + className + 
+                type switch
+            {
+                SynthesizedInterfaceType.Default => "Class",
+                SynthesizedInterfaceType.Factory => "Factory",
+                SynthesizedInterfaceType.Static => "Static",
+                _ => "",
+            };
+        }
+
+        void AddSynthesizedInterfaces(TypeDeclaration classDeclaration)
+        {
+            HashSet<ISymbol> classMembersFromInterfaces = new HashSet<ISymbol>();
+            INamedTypeSymbol classSymbol = classDeclaration.Node as INamedTypeSymbol;
+            foreach (var @interface in classSymbol.AllInterfaces)
+            {
+                foreach(var interfaceMember in @interface.GetMembers())
+                {
+                    classMembersFromInterfaces.Add(classSymbol.FindImplementationForInterfaceMember(interfaceMember));
+                }
+            }
+
+            AddSynthesizedInterface(
+                classDeclaration,
+                SynthesizedInterfaceType.Static,
+                classMembersFromInterfaces);
+
+            AddSynthesizedInterface(
+                classDeclaration,
+                SynthesizedInterfaceType.Factory,
+                classMembersFromInterfaces);
+
+            AddSynthesizedInterface(
+                classDeclaration,
+                SynthesizedInterfaceType.Default,
+                classMembersFromInterfaces);
+        }

```

Comment:
overridable and composable?

Reply:
on the todo, will be added once I got the cswinrt side working for basic components.  Will add a todo comment.